File: src/main/java/com/example/config/ModConfig.java
package com.example.config;

import com.example.TemplateMod;
import com.example.config.categories.SimpleCategory;
import com.google.gson.GsonBuilder;
import dev.isxander.yacl3.api.ConfigCategory;
import dev.isxander.yacl3.api.YetAnotherConfigLib;
import dev.isxander.yacl3.config.v2.api.ConfigClassHandler;
import dev.isxander.yacl3.config.v2.api.SerialEntry;
import dev.isxander.yacl3.config.v2.api.serializer.GsonConfigSerializerBuilder;
import dev.isxander.yacl3.platform.YACLPlatform;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.text.Text;

import java.util.ArrayList;
import java.util.List;

public class ModConfig {
    public static final ConfigClassHandler<ModConfig> HANDLER = ConfigClassHandler.createBuilder(ModConfig.class)
            .serializer(config -> GsonConfigSerializerBuilder.create(config)
                    .setPath(YACLPlatform.getConfigDir().resolve(TemplateMod.MOD_ID+".json5"))
                    .appendGsonBuilder(GsonBuilder::setPrettyPrinting)
                    .setJson5(true)
                    .build())
            .build();

    public static List<ConfigCategory> CATEGORIES = new ArrayList<>();

    @SerialEntry
    public boolean myCoolBoolean = true;

    public static Screen getConfigScreen(Screen parent) {
        return YetAnotherConfigLib.createBuilder()
                .title(Text.literal("Template Mod Config"))
                .categories(CATEGORIES)
                .build()
                .generateScreen(parent);
    }

    public static void initialize() {
        CATEGORIES.add(SimpleCategory.get());
        load();
    }

    public static ModConfig get() {
        return HANDLER.instance();
    }

    public static void save() {
        HANDLER.save();
    }

    public static void load() {
        HANDLER.load();
    }
}

File: src/main/kotlin/com/example/config/ModConfig.kt
package com.example.config

import com.example.TemplateMod
import com.mojang.serialization.Codec
import dev.isxander.yacl3.config.v3.JsonFileCodecConfig
import dev.isxander.yacl3.config.v3.register
import dev.isxander.yacl3.platform.YACLPlatform

open class ModConfig : JsonFileCodecConfig<ModConfig>(
    YACLPlatform.getConfigDir().resolve(TemplateMod.MOD_ID + ".json")
) {

    val myCoolBoolean by register<Boolean>(true, Codec.BOOL)

    companion object : ModConfig() {
        init {
            if (!loadFromFile()) {
                saveToFile()
            }
        }
    }
}
